lib/puppet/pops/types/iterable.rb

```diff
@@ -0,0 +1,291 @@
+      iter
+    end
+
+    # Produces an `Iterable` for one of the following types with the following characterstics:
+    #
+    # `String`       - yields each character in the string
+    # `Array`        - yields each element in the array
+    # `Hash`         - yields each key/value pair as a two element array
+    # `Integer`      - when positive, yields each value from zero to the given number
+    # `PIntegerType` - yields each element from min to max (inclusive) provided min < max and neither is unbounded.
+    # `PEnumtype`    - yields each possible value of the enum.
+    # `Range`        - yields an iterator for all elements in the range provided that the range start and end
+    #                  are both integers or both strings and start is less than end using natural ordering.
+    # `Dir`          - yields each name in the directory
+    #
+    # The value `nil` is returned for all other objects.
+    #
+    # @param o [Object] The object to produce an `Iterable` for
+    # @return [Iterable,nil] The produced `Iterable` or `nil` if it couldn't be produced
+    #
+    # @api public
+    def self.on(o)
+      case o
+      when Iterable
+        o
+      when String
+        Iterator.new(PStringType.new(PIntegerType.new(1, 1)), o.each_char)
+      when Array
+        tc = TypeCalculator.singleton
+        Iterator.new(tc.unwrap_single_variant(PVariantType.new(o.map {|e| tc.infer_set(e) })), o.each)
+      when Hash
+        # Each element is a two element [key, value] tuple.
+        tc = TypeCalculator.singleton
+        Iterator.new(PTupleType.new([
+          tc.unwrap_single_variant(PVariantType.new(o.keys.map {|e| tc.infer_set(e) })),
+          tc.unwrap_single_variant(PVariantType.new(o.values.map {|e| tc.infer_set(e) }))], PIntegerType.new(2,2)), o.each_pair)
+      when Integer
+        o > 0 ? IntegerRangeIterator.new(PIntegerType.new(0, o - 1)) : nil
+      when PIntegerType
+        o.enumerable? ? IntegerRangeIterator.new(o) : nil
+      when PEnumType
+        Iterator.new(o, o.values)
+      when Range
+        min = o.min
+        max = o.max
+        if min.is_a?(Integer) && max.is_a?(Integer) && max > min
+          IntegerRangeIterator.new(PIntegerType.new(min, max))
+        elsif min.is_a?(String) && max.is_a?(String) && max > min
+          # A generalized element type where only the size is inferred is used here since inferring the full
+          # range might waste a lot of memory.
+          if min.length < max.length
+            shortest = min
+            longest = max
+          else
+            shortest = max
+            longest = min
+          end
+          Iterator.new(PStringType.new(PIntegerType.new(shortest.length, longest.length)), o.each)
+        else
+          # Unsupported range. It's either descending or nonsensical for other reasons (float, mixed types, etc.)
+          nil
+        end
+      when Dir
```

Comment:
It was the only implementation in Ruby that we doesn't otherwise cover that seemed to make sense. I can remove it.